src/bin/nydusd/fusedev.rs

```diff
@@ -229,18 +230,22 @@ impl NydusDaemon for FusedevDaemon {
     }
 
     fn wait(&self) -> DaemonResult<()> {
-        let mut guard = self.threads.lock().unwrap();
-
-        while let Some(handle) = guard.pop() {
-            handle
-                .join()
-                .map_err(|e| {
-                    DaemonError::WaitDaemon(
-                        *e.downcast::<Error>()
-                            .unwrap_or_else(|e| Box::new(eother!(e))),
-                    )
-                })?
-                .map_err(DaemonError::WaitDaemon)?;
+        loop {
+            let handle = self.threads.lock().unwrap().pop();
```

Comment:
This changes where threads.lock protects,  it'll unlock  `threads` after pop(), why is that?

Reply:
If we block on `handle.join` with the lock at hand, we cannot add more threads to the waiting list, which is necessary for the upgrade use cases, where we just start the state machine thread and then start more fuse service threads later on.

Reply:
Please correct me if I'm wrong.

In case of upgrade usecase, `d.wait()` is called by old binary's daemon right before old binary quits itself, so why would be more fuse service threads coming in to serve old binary?

With that being said, it looks like early unlock is fine because there is unlikely a race here.

Reply:
It is not about the old binary's daemon. The new daemon will start and block on waiting for the state machine thread. And then the control path manager will send fd to the new daemon and ask it to take over the service. Then the new daemon will need to start fuse server threads and add them to the waiting list. That's why we need to drop the lock here.